tfhe/src/high_level_api/details.rs

```diff
@@ -212,6 +214,82 @@ macro_rules! define_key_structs {
+
+            #[derive(Clone, Debug, Default, ::serde::Deserialize, ::serde::Serialize)]
+            pub(crate) struct [<$base_struct_name KeySwitchingParameters>] {
+                $(
+                    pub(crate) [<$name _params>]: Option<[<$base_ty_name KeySwitchingParameters>]>,
+                )*
+            }
+
+            #[derive(Clone, Debug, ::serde::Deserialize, ::serde::Serialize)]
+            pub(crate) struct [<$base_struct_name KeySwitchingKey>] {
+                $(
+                    pub(super) [<$name _key>]: Option<[<$base_ty_name KeySwitchingKey>]>,
+                )*
+            }
+
+            impl [<$base_struct_name KeySwitchingKey>] {
+                pub(crate) fn new(
+                    key_pair_1: (&[<$base_struct_name ClientKey>], &[<$base_struct_name ServerKey>]),
+                    key_pair_2: (&[<$base_struct_name ClientKey>], &[<$base_struct_name ServerKey>]),
+                    parameters: [<$base_struct_name KeySwitchingParameters>],
+                ) -> Self {
+                    Self {
+                        $(
+                            [<$name _key>]: match (
+                                key_pair_1.0.[<$name _key>].as_ref(),
+                                key_pair_1.1.[<$name _key>].as_ref(),
+                                key_pair_2.0.[<$name _key>].as_ref(),
+                                key_pair_2.1.[<$name _key>].as_ref(),
+                                parameters.[<$name _params>],
+                            ) {
+                                 (Some(ck1), Some(sk1), Some(ck2), Some(sk2), Some(ksk_params)) => {
+                                    Some(<[<$base_ty_name KeySwitchingKey>]>::new((ck1, sk1), (ck2, sk2), ksk_params))
+                                },
+                                _ => None
+                            },
+                        )*
+                    }
+                }
+
+                pub(crate) fn for_same_parameters(
+                    key_pair_1: (&[<$base_struct_name ClientKey>], &[<$base_struct_name ServerKey>]),
+                    key_pair_2: (&[<$base_struct_name ClientKey>], &[<$base_struct_name ServerKey>]))
+                -> Option<Self> {
+                    Some(Self {
+                        $(
+                            [<$name _key>]: match (
+                                key_pair_1.0.[<$name _key>].as_ref(),
+                                key_pair_1.1.[<$name _key>].as_ref(),
+                                key_pair_2.0.[<$name _key>].as_ref(),
+                                key_pair_2.1.[<$name _key>].as_ref())
+                            {
+                                (Some(ck1), Some(sk1), Some(ck2), Some(sk2)) => {
+                                    // Forward error if parameters are not compatible
+                                    let key = <[<$base_ty_name KeySwitchingKey>]>::for_same_parameters((ck1, sk1), (ck2, sk2))?;
+                                    Some(key)
```

Comment:
no need to forward ? so need for the `?` and the let binding here I would say as it already returns an option

Reply:
ermf, here its a bit more complex than that I think:

If we don't have a `Some(_)` for each key in the pairs the match forwards a None the be placed at the member level 
```rust
e.g: Self {
uint2: None
}
```

If we do have all Some(_), then we try to create a ksk using the `for_same_param`, however that function will return None if the keys in the pair do not actually use the same params.

In the case they indeed do not use the same params, we want the whole function to return None, and not just the key member (eg uint2_key) to be None.

Reply:
if it works like this fine, was just thinking it looked odd

Reply:
One potential thing to make it clearer would be to create an error type

```rust
struct IncompatibleParameters(type); // type is an enum variant, either FheBool, FheUint2, Integer

and have the `for_same_params` functions return a `Result<Self, IncompatibleParameter>`
```

Reply:
I think it derserves a comment though potentially

Reply:
Ok i'll add the comment, and not the error struct as its les tedious